src/borrow_tracker/tree_borrows/perms.rs

```diff
@@ -339,30 +341,37 @@ pub mod diagnostics {
-                        // No transition has `Disabled` as its `.from` unless it's a noop.
-                        (_, Disabled, _) => unreachable!("noop transition"),
-                        // These imply that the permission became `Disabled`
-                        // (resp. `Frozen`, `Active`) and was later found `Reserved | Active | Frozen`
-                        // (resp. `Reserved | Active`, `Reserved`), which is impossible because
-                        // permissions only evolve in the order `Reserved -> Active -> Frozen -> Disabled`.
-                        (Disabled, Reserved { .. } | Active | Frozen, _)
-                        | (Frozen, Reserved { .. } | Active, _)
-                        | (Active, Reserved { .. }, _) =>
-                            unreachable!("violates monotony of permissions"),
+                        // We absolutely want to know where it was activated.
+                        (Active, Active, Frozen | Disabled) => true,
+                        // And knowing where it became Frozen is also important.
+                        (Frozen, Frozen, Disabled) => true,
                         // If the error is a transition `Frozen -> Disabled`, then we don't really
                         // care whether before that was `Reserved -> Active -> Frozen` or
                         // `Reserved -> Frozen` or even `Frozen` directly.
                         // The error will only show either
                         // - created as Frozen, then Frozen -> Disabled is forbidden
                         // - created as Reserved, later became Frozen, then Frozen -> Disabled is forbidden
                         // In both cases the `Reserved -> Active` part is inexistant or irrelevant.
-                        (Active, Frozen, _) => false,
+                        (Active, Frozen, Disabled) => false,
+
+                        // `Reserved -> Frozen` does not trigger protectors.
+                        (_, Reserved { .. }, Frozen) => unreachable!("cannot cause an error"),
+                        // No transition has `Reserved` as its `.to` unless it's a noop.
+                        (Reserved { .. }, _, _) => unreachable!("self is a noop transition"),
+                        (_, Disabled, Disabled) | (_, Frozen, Frozen) | (_, Active, Active) =>
+                            unreachable!("err contains a noop transition"),
+
+                        // Permissions only evolve in the order `Reserved -> Active -> Frozen -> Disabled`,
+                        // so permissions found must be increasing in the order
+                        // `self.from < self.to <= forbidden.from < forbidden.to`.
+                        (Disabled, Reserved { .. } | Active | Frozen, _)
+                        | (Frozen, Reserved { .. } | Active, _)
+                        | (Active, Reserved { .. }, _) =>
+                            unreachable!("violates monotony of permissions between self and err"),
```

Comment:
What is "monotony of permissions"? I think "monotonicity" would be grammatically correct but that is still a strange use of the term I think. This is just describing the allowed transitions between states, right?